### PR TITLE
[BIOMAGE-1669] Add the ui features to create cellset based on expression

### PR DESCRIPTION
**Pull request by StefanBabukov**
# Description

# Details
#### URL to issue
https://biomage.atlassian.net/browse/BIOMAGE-1669

#### Link to staging deployment URL (or set N/A)
https://ui-grasp6-ui625.scp-staging.biomage.net/

#### Integration test branch
master

# Merge checklist
Your changes will be ready for merging after all of the steps below have been completed.

### Code updates
Have best practices and ongoing refactors being observed in this PR
- [ ] Migrated any selector / reducer used to the new format.

### Manual/unit testing
- [ ] Tested changes using InfraMock locally **or** no tests required for change, e.g. Kubernetes chart updates.
- [ ] Validated that current unit tests for code work as expected and are sufficient for code coverage **or** no unit tests required for change, e.g. documentation update.
- [ ] Unit tests written **or** no unit tests required for change, e.g. documentation update.

### Integration testing
**You must check the box below** to run integration tests on the latest commit on your PR branch.
Integration tests have to pass before the PR can be merged. Without checking the box, your PR
**will not pass** the required status checks for merging.

- [ ] Started end-to-end tests on the latest commit.

### Documentation updates
- [ ] Relevant Github READMEs updated **or** no GitHub README updates required.
- [ ] Relevant Wiki pages created/updated **or** no Wiki updates required.

### Optional
- [ ] Staging environment is unstaged before merging.
- [ ] Photo of a cute animal attached to this PR.

### CODE REVIEW

**Comment by StefanBabukov** [comment]
@vickymorrison  What should the max value be for the gene expression?

**Comment by vickymorrison** [comment]
About your question `What should the max value be for the gene expression?` -> I have no idea.
@ogibson-biomage Can you help?

File: src/__test__/redux/actions/componentConfig/__snapshots__/loadPlotConfig.test.js.snap [hunk 1]
```diff
@@ -14,7 +14,6 @@ Object {
       "xAxisText": "#UMIs in cell",
       "yAxisText": "#UMIs * #Cells",
     },
-    "binStep": 200,
```
Comment (StefanBabukov) [inline src/__test__/redux/actions/componentConfig/__snapshots__/loadPlotConfig.test.js.snap:17]: These binStep removals are related to this discussion https://this-is-biomage.slack.com/archives/C015CHSUDRU/p1643803335465179

File: src/__test__/components/data-exploration/generic-gene-table/SelectionActions.test.jsx [hunk 2]
```diff
@@ -5,6 +5,7 @@ import { Button, Typography } from 'antd';
 import { Provider } from 'react-redux';
 import thunk from 'redux-thunk';
 import configureMockStore from 'redux-mock-store';
+import CreateCellSetModal from 'components/data-exploration/generic-gene-table/CreateCellSetModal';
 import SelectionActions from '../../../../components/data-exploration/generic-gene-table/SelectionActions';
 import { GENES_DESELECT } from '../../../../redux/actionTypes/genes';
```
Comment (aerlaut) [inline src/__test__/components/data-exploration/generic-gene-table/SelectionActions.test.jsx]: Can you change these paths to use absolute paths?

File: src/components/data-exploration/generic-gene-table/SelectionActions.jsx [hunk 3]
```diff
@@ -10,10 +10,10 @@ import { CheckCircleOutlined } from '@ant-design/icons';
 import { CopyToClipboard } from 'react-copy-to-clipboard';
 import { changeGeneSelection } from 'redux/actions/genes';
 import GeneSelectionStatus from 'redux/actions/genes/geneSelectionStatus';
-
 import ComponentActions from 'components/data-exploration/generic-gene-table/ComponentActions';
 
 import { COMPONENT_TYPE } from 'components/data-exploration/heatmap/HeatmapPlot';
+import CreateCellSetModal from './CreateCellSetModal';
```
Comment (aerlaut) [inline src/components/data-exploration/generic-gene-table/SelectionActions.jsx]: Can you use absolute import for this?

File: src/components/data-exploration/generic-gene-table/SelectionActions.jsx [hunk 4]
```diff
@@ -97,6 +98,20 @@ const SelectionActions = (props) => {
             {listed ? 'Hide' : 'List'}
           </Button>
           <ComponentActions name='Heatmap' experimentId={experimentId} componentType={COMPONENT_TYPE} />
+          <Button
+            type='link'
+            onClick={() => setCreateCellSetModalVisible(!createCellSetModalVisible)}
```
Comment (aerlaut) [inline src/components/data-exploration/generic-gene-table/SelectionActions.jsx]: can you add `size='small'` here so that it's the same size with the other links?

File: src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        onFinish={({ filterForm }) => createCellSet(filterForm)}
+        size='middle'
+      >
+        <Form.List
+          name='filterForm'
+          initialValue={intialFormValues}
+        >
+          {(fields) => (
+            <>
+              <Row style={{ overflow: 'scroll', height: formHeight }}>
```
Comment (aerlaut) [inline src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx]: can you use `overflow-y:auto` instead, so the scroll bar will not appear if it doesn't need to appear?

File: src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+  const comparisonOptions = [{
+    value: 'greaterThan',
+    label: 'Greater than',
+  }, {
+    value: 'lessThan',
+    label: 'Less than',
+  },
+  ];
+
+  const formHeight = selectedGenes.length < 5 ? selectedGenes.length * 75 : 300;
```
Comment (aerlaut) [inline src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx]: why 75 per gene and limit the height to 300 (4 genes)? I see that there are still a lot of spaces that we can use. Aside from that, can you also reduce the size of the text and the spacing between each genes?

image

File: src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+  return (
+    <Modal
+      visible
+      title='Create a new cell set based on gene expression'
+      onCancel={onCancel}
+      footer={null}
+      width='530px'
```
Comment (aerlaut) [inline src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx]: Why do we need to fix the width? do you mean `min-width`?

File: src/components/data-exploration/generic-gene-table/SelectionActions.jsx (hunk 4 above)
Comment (StefanBabukov) [reply src/components/data-exploration/generic-gene-table/SelectionActions.jsx]: small is by default so we dont really need it

File: src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx (hunk 6 above)
Comment (StefanBabukov) [reply src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx]: I increased it because there are some long genes which dont fit in one line

Comment (StefanBabukov) [reply src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx]: if we want all the rows to be aligned

Comment (StefanBabukov) [reply src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx]: Other than that I wasnt sure how many genes should be shown before the scroll

File: src/components/data-exploration/generic-gene-table/SelectionActions.jsx (hunk 4 above)
Comment (aerlaut) [reply src/components/data-exploration/generic-gene-table/SelectionActions.jsx]: No, the default is `medium`([reference](https://ant.design/components/button/#API)), that's why it's showing up on a different size than others 
image

File: src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx (hunk 6 above)
Comment (aerlaut) [reply src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx]: Try not to fix the `width`, so the gene names would fit on one line?

File: src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+                          max={100}
+                          placeholder='Insert value'
+                        />
+                      </Form.Item>
+                    </Space>
+                  );
+                })}
+              </Row>
+              <Alert type='info' message='To edit the list of genes, return to the gene list and change your selection.' />
+              <Divider style={{ marginBottom: '10px' }} />
```
Comment (aerlaut) [inline src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx]: Instead of using divider, can you put this button int the modal's footer instead?

File: src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx (hunk 6 above)
Comment (StefanBabukov) [reply src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx]: Yeah this way the rows wont be aligned though?

Comment (StefanBabukov) [reply src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx]: image

Comment (StefanBabukov) [reply src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx]: My OCD didnt let me go with that :D

Comment (StefanBabukov) [reply src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx]: @vickymorrison  what do you think about this?

Comment (aerlaut) [reply src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx]: Maybe you can try something like 

    <Row justify="end">
      <Col flex="auto">Fill Rest</Col>
      <Col flex="100px">This is the comparison dropdown</Col>
      <Col flex="100px">This is the input field</Col>
    </Row>

Does that work?

Comment (StefanBabukov) [reply src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx]: ill try it i always forget these `Row` and `Col` exist

**Comment by ivababukova** [comment]
When I click on the `Cellset` button, these 2 red warnings appear in the console, can you fix them:

Screenshot 2022-02-08 at 10 15 36
Screenshot 2022-02-08 at 10 15 28

**Comment by ivababukova** [comment]
The actual functionality of this will not work until we do the other tickets in this epic, so don't forget just before merging to either hide this button, or disable it and add a tooltip that this functionality is going to be supported soon

**Comment by ivababukova** [comment]
You've forgotten to add the cancel button (see the wireframe: https://docs.google.com/presentation/d/1Kz5jlcJPMAfLed8L4CmosceV7milSdIqhh9Jcy_KQ8A/edit#slide=id.g10e6a0da698_0_5)

File: src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    label: 'Less than',
+  },
+  ];
+
+  return (
+    <Modal
+      visible
+      title='Create a new cell set based on gene expression'
+      onCancel={onCancel}
+      footer={[<Button type='primary'>Create</Button>]}
```
Comment (ivababukova) [inline src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx]: Nothing happens when I click "Create", is this intentional? Are we planning to implement the functionality of create after the backend is merged?

**Comment by ivababukova** [comment]
@vickymorrison this `Cellset` button will appear on the differential expr results table, this is intentional, right?

File: src/components/data-exploration/generic-gene-table/SelectionActions.jsx [hunk 10]
```diff
@@ -97,6 +98,21 @@ const SelectionActions = (props) => {
...
+            Cellset
+          </Button>
+          {
+            createCellSetModalVisible && (
+              <CreateCellSetModal
+                selectedGenes={selectedGenes}
+                onCancel={() => setCreateCellSetModalVisible(false)}
+              />
+            )
+          }
```
Comment (ivababukova) [inline src/components/data-exploration/generic-gene-table/SelectionActions.jsx]: I wouldn't have this code in `SelectionActions` - it is not the job of `SelectionActions` to know when the Cellset button should be visible or not. This logic should be handled inside `CreateCellSetModal`. Also, this `setCreateCellSetModalVisible` variable is a duplicate of `selectedGenes.length !== 0`, which is checked at the beginning of the return statement in `SelectionActions`.
You can delete this `createCellSetModalVisible` state variable and follow the concept behind the implementation of of `ComponentActions`: have the `CreateCellSetModal` just called there and later on handle the cellset creation logic inside `CreateCellSetModal`.

File: src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+                      </Form.Item>
+                    </Col>
+                    <Col>
+                      <Form.Item
+                        name={[field.name, 'comparison']}
+                        style={formItemStyle}
+                      >
+                        <Select
+                          placeholder='Select comparison'
+                          defaultValue='greaterThan'
```
Comment (ivababukova) [inline src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx]: instead of hardcoding `greaterThan`, you can choose the first element out of `comparisonOptions`

File: src/__test__/components/data-exploration/generic-gene-table/CreateCellSetModal.test.jsx [hunk 12]
```diff
@@ -0,0 +1,20 @@
+import React from 'react';
+import { render, screen } from '@testing-library/react';
+
+import CreateCellSetModal from 'components/data-exploration/generic-gene-table/CreateCellSetModal';
+
+describe('Create cell set modal', () => {
+  it('renders correctly', () => {
```
Comment (ivababukova) [inline src/__test__/components/data-exploration/generic-gene-table/CreateCellSetModal.test.jsx:25]: what about a test that gets closed on cancel or on clicking the cross?

File: src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+  Modal, Form, Button, Select, InputNumber, Row, Alert, Col,
+} from 'antd';
+import PropTypes from 'prop-types';
+
+const formItemStyle = { margin: '0.375rem 0' };
+
+const CreateCellSetModal = (props) => {
+  const { selectedGenes, onCancel } = props;
+  const [form] = Form.useForm();
+  const intialFormValues = selectedGenes.map((gene) => ({ selectedGenes: gene }));
```
Comment (ivababukova) [inline src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx]: The `CreateCellSetModal` will need to be tied up to the redux store later on anyway to implement the calls to create cellset actions, so there is no point passing `selectedGenes` as props and avoiding Redux right now. In this case, you can have `  const selectedGenes = useSelector((state) => state.genes.selected);` and remove variable `intialFormValues`, just use `selectedGenes`

Comment (ivababukova) [reply src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx]: Also, why are you representing `initialFormValues` as an array of objects like this:
`[ {selectedGenes: 'CD3D'},  {selectedGenes: 'PPBP'},  {selectedGenes: 'Chil3'}]`? Why not representing it like this:
`['CD3D', 'PPBP', 'Chil3']`?

File: src/__test__/components/data-exploration/generic-gene-table/CreateCellSetModal.test.jsx (hunk 12 above)
Comment (StefanBabukov) [reply src/__test__/components/data-exploration/generic-gene-table/CreateCellSetModal.test.jsx:25]: There is a test in SelectionActions

File: src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx (hunk 9 above)
Comment (StefanBabukov) [reply src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx]: Yes

File: src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx (hunk 13 above)
Comment (StefanBabukov) [reply src/components/data-exploration/generic-gene-table/CreateCellSetModal.jsx]: because the fields are assigned to the corresponding form.item name  - `selectedGenes` is the first text field

File: src/redux/reducers/componentConfig/initialState.js [hunk 14]
```diff
@@ -512,7 +510,6 @@ const mitochondrialFractionLogHistogram = {
   colour: colourBaseState,
   marker: markerBaseState,
   labels: labelBaseState,
-  binStep: 0.05,
```
Comment (ivababukova) [inline src/redux/reducers/componentConfig/initialState.js:515]: why did you delete these?

File: src/__test__/components/data-exploration/generic-gene-table/SelectionActions.test.jsx [hunk 15]
```diff
@@ -206,4 +175,33 @@ describe('SelectionIndicator', () => {
...
+  //   expect(component.find(CreateCellSetModal).length).toEqual(1);
+
+  //   // check if the genes are passed
+  //   expect(component.find(CreateCellSetModal).prop('selectedGenes'))
+  //      .toStrictEqual(['CEMIP', 'TIMP3']);
+
+  //   const closeButton = component.find('.ant-modal-close');
+  //   closeButton.simulate('click');
+  //   expect(component.find(CreateCellSetModal).length).toEqual(0);
+  // });
```
Comment (ivababukova) [inline src/__test__/components/data-exploration/generic-gene-table/SelectionActions.test.jsx]: Commented tests is not a good idea -- we can easily forget about them and never change them. Better to uncomment them and change them to test for the behaviour that you are committing to master now. Later on, when we change the component fucntionality, we will adapt the tests.

File: src/redux/reducers/componentConfig/initialState.js (hunk 14 above)
Comment (StefanBabukov) [reply src/redux/reducers/componentConfig/initialState.js:515]: These binStep removals are related to this discussion https://this-is-biomage.slack.com/archives/C015CHSUDRU/p1643803335465179

**Review by ivababukova — APPROVED** [review]
